### PR TITLE
Add support for FixedSizeList type in `arrow_cast`, hashing

### LINKED ISSUE #8343: Support `FixedSizeList` for `arrow_cast`
### Is your feature request related to a problem or challenge?

In the next arrow-rs(50.0), it will support the cast between `FixedSizeList` / 'List' / 'LargeList'. 

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
## Which issue does this PR close?

Closes #8343

## Rationale for this change
- add hash function for 'FixedSizeList'
- support cast between `FixedSizeList` and `List` / `LargeList`

### CODE REVIEW

**Comment by Weijun-H** [comment]
stall until the next arrow-rs (50.0)

**Review by alamb — COMMENTED** [review]
I wonder why this PR is waiting for the next arrow release (arrow `50.0.0`?)

**Comment by Weijun-H** [comment]
> I wonder why this PR is waiting for the next arrow release (arrow `50.0.0`?)

Because it just update the support  cast from List/LargeList to FixedList @alamb 
https://github.com/apache/arrow-rs/blob/df69ef57d055453c399fa925ad315d19211d7ab2/arrow-cast/src/cast.rs#L808-L815

**Comment by alamb** [comment]
I see -- so while this code doesn't directly depend on the arrow release, it won't be very helpful until that support released. Makes sense to me. Thank you @Weijun-H

**Comment by alamb** [comment]
❤️

File: datafusion/sqllogictest/test_files/arrow_typeof.slt [hunk 1]
```diff
@@ -384,4 +384,35 @@ LargeList(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, di
...
+LargeList(Field { name: "item", data_type: LargeList(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })
+
+## FixedSizeList
+
+query ?
+select arrow_cast(null, 'FixedSizeList(1, Int64)');
+----
+NULL
+
+#TODO: arrow-rs doesn't support it yet
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/arrow_typeof.slt:396]: Shouldn't we be casting `[1]` (not `1`)?

File: datafusion/common/src/hash_utils.rs [hunk 2]
```diff
@@ -267,6 +268,38 @@ where
     Ok(())
 }
 
+fn hash_fixed_list_array(
```
Comment (alamb) [inline datafusion/common/src/hash_utils.rs:271]: I didn't see any test coverage for this new code -- e.g. either unit tests for hashing or a higher level test like `GROUP BY <FixedListArray>`

Can you either ensure this code is tested somehow, or else perhaps move the hash support to a different PR so we can merge the `arrow_cast` support ?

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H  -- as always, your contributions are most appreciated

File: datafusion/sqllogictest/test_files/arrow_typeof.slt (hunk 1 above)
Comment (Weijun-H) [reply datafusion/sqllogictest/test_files/arrow_typeof.slt:396]: I noticed that the List supports casting from UTF8 to List with a single size. Therefore, I think FixedSizeList should also support it. 
```
select arrow_cast('1', 'LargeList(Int64)');
----
[1]
```

File: datafusion/common/src/hash_utils.rs [hunk 3]
```diff
@@ -546,6 +583,30 @@ mod tests {
...
+                data, 3,
+            )) as ArrayRef;
+        let random_state = RandomState::with_seeds(0, 0, 0, 0);
+        let mut hashes = vec![0; list_array.len()];
+        create_hashes(&[list_array], &random_state, &mut hashes).unwrap();
+        assert_eq!(hashes[0], hashes[5]);
+        assert_eq!(hashes[1], hashes[4]);
+        assert_eq!(hashes[2], hashes[3]);
+    }
+
```
Comment (Weijun-H) [inline datafusion/common/src/hash_utils.rs:609]: a unit test for the hash function.

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @Weijun-H
